browser/brave_ads/brave_ads_host.cc

```diff
@@ -0,0 +1,162 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/brave_ads/brave_ads_host.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "brave/browser/brave_ads/ads_service_factory.h"
+#include "brave/browser/brave_rewards/rewards_service_factory.h"
+#include "brave/browser/extensions/brave_component_loader.h"
+#include "brave/components/brave_ads/browser/ads_service.h"
+#include "brave/components/brave_rewards/browser/rewards_service.h"
+#include "chrome/browser/extensions/extension_service.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "content/public/browser/web_contents.h"
+#include "extensions/browser/extension_system.h"
+#include "extensions/common/constants.h"
+
+namespace brave_ads {
+
+namespace {
+
+constexpr char kAdsEnableRelativeUrl[] = "request_ads_enabled_panel.html";
+
+}  // namespace
+
+BraveAdsHost::BraveAdsHost(content::WebContents* web_contents)
+    : WebContentsObserver(web_contents) {
+  DCHECK(web_contents);
+}
+
+BraveAdsHost::~BraveAdsHost() {
+  DCHECK(!callback_);
+}
+
+void BraveAdsHost::RequestAdsEnabled(RequestAdsEnabledCallback callback) {
+  DCHECK(callback);
+
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  DCHECK(profile);
+
+  const AdsService* ads_service = AdsServiceFactory::GetForProfile(profile);
+  brave_rewards::RewardsService* rewards_service =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile);
+  if (!rewards_service || !ads_service || !ads_service->IsSupportedLocale()) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  if (ads_service->IsEnabled()) {
+    std::move(callback).Run(true);
+    return;
+  }
+
+  if (callback_) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  Browser* browser = chrome::FindBrowserWithWebContents(web_contents());
+  if (!browser) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  callback_ = std::move(callback);
+
+  rewards_service->StartProcess(base::DoNothing());
```

Comment:
We shouldn't be starting the process and not sure why that's a public method in the first place. rewards/ads should be enabled by toggling the prefs prefs::kAutoContributeEnabled/ads::prefs::kEnabled

Reply:
Just curious whether we have any plans to fix the weirdness of enabling the service via toggling a pref?... It's one of the most confusing things I've ever encountered in our codebase

Reply:
not sure what you mean, all kinds of things get enabled by setting prefs, but it's actually apparently quite a bit more complicated and confusing at the moment

Reply:
> We shouldn't be starting the process and not sure why that's a public method in the first place. rewards/ads should be enabled by toggling the prefs prefs::kAutoContributeEnabled/ads::prefs::kEnabled

Seems like this call is not needed right now, because current popup doesn't depend on RewardsExtension.State. I will remove this call:

`rewards_service->StartProcess(base::DoNothing());`